storage/innobase/os/os0file.cc

```diff
@@ -9615,20 +9610,44 @@ byte *Encryption::encrypt(const IORequest &type, byte *src, ulint src_len,
 @return DB_SUCCESS or error code */
 dberr_t Encryption::decrypt_log_block(const IORequest &type, byte *src,
                                       byte *dst) {
-  ulint data_len;
   ulint main_len;
   ulint remain_len;
   byte remain_buf[MY_AES_BLOCK_SIZE * 2];
   byte *ptr = src;
 
   /* This is data size which need to encrypt. */
-  data_len = OS_FILE_LOG_BLOCK_SIZE - LOG_BLOCK_HDR_SIZE;
+  const ulint unencrypted_trailer_size =
+      (m_type == Encryption::KEYRING) ? LOG_BLOCK_TRL_SIZE : 0;
+  const ulint data_len =
+      OS_FILE_LOG_BLOCK_SIZE - LOG_BLOCK_HDR_SIZE - unencrypted_trailer_size;
+
   main_len = (data_len / MY_AES_BLOCK_SIZE) * MY_AES_BLOCK_SIZE;
   remain_len = data_len - main_len;
 
   ptr += LOG_BLOCK_HDR_SIZE;
   switch (m_type) {
+    case Encryption::KEYRING: {
+      const ulint block_crc = log_block_calc_checksum_crc32(src);
+      const ulint written_crc = log_block_get_checksum(src);
+
+      const ulint enc_key_version = written_crc - block_crc;
+
+      if (m_key_version != enc_key_version &&
+          enc_key_version != REDO_LOG_ENCRYPT_NO_VERSION) {
+        redo_log_key *mkey = redo_log_key_mgr.load_key_version(enc_key_version);
+        m_key_version = mkey->version;
+        m_key = reinterpret_cast<unsigned char *>(mkey->key);
+      }
+#ifdef UNIV_ENCRYPT_DEBUG
+      fprintf(stderr, "Decrypting with key version: %lu", enc_key_version);
```

Comment:
These are too generic messages even for debug. Suggest to move ib::info() and avoid format specifier